src/App.tsx

```diff
@@ -11,29 +12,73 @@ import { OnboardContext } from "common/context/OnboardContext";
 import { recoverPublicKey } from "./features/vote/helpers/recoverPublicKey";
 import { derivePrivateKey } from "./features/vote/helpers/derivePrivateKey";
 
+import currentSigningMessage from "common/currentSigningMessage";
+import { useCurrentRoundId } from "hooks";
+
 interface Props {
   queryClient: QueryClient;
 }
 
-export interface SigningKeys {
+export interface SigningKey {
   [key: string]: {
     publicKey: string;
     privateKey: string;
   };
 }
 
+export interface SigningKeys {
+  [key: string]: SigningKey;
+}
+
+const SIGNING_KEYS_STORAGE_KEY = "signingKeys";
+
 function App(props: Props) {
   const [signingKeys, setSigningKeys] = useState<SigningKeys>({});
 
   const { state, disconnect, dispatch } = useContext(OnboardContext);
   const { error, removeError, addError } = useContext(ErrorContext);
+  const { data: currentRoundId } = useCurrentRoundId();
 
   useEffect(() => {
     if (state.signer && state.address) {
       const address = state.address;
-      const message = "Login to UMA Voter dApp";
-      const keyExists = signingKeys[address];
-      if (!keyExists) {
+      const message = currentSigningMessage(Number(currentRoundId));
+      const hashedMessage = ethers.utils.formatBytes32String(message);
+      const keysString = localStorage.getItem(SIGNING_KEYS_STORAGE_KEY);
+      if (keysString) {
+        const keys = JSON.parse(keysString) as SigningKeys;
+        const keyExists = keys[hashedMessage][address];
```

Comment:
Thoughts on putting some try-catches around this logic? Seems like we'd want an escape hatch if anything fails so we always fall back to just re-requesting the signature. Thoughts? Or do you think that could cause issues?

Total side note: as a long-term security measure to avoid someone somehow corrupting local storage to break a voter's reveal (however unlikely that may be), we could add a step where we validate the private key -- we could do this similarly to how we verify signatures on-chain for the reveal snapshot. How likely do you think it is that someone is able to corrupt local storage? Could a scammy extension do this?

Reply:
Yeah the issue with your security hole is a problem. We are only looking if a signing key exists -- if someone got access to the localStorage they could scramble the value. The only fix would be to re-sign, which they could by clearing localstorage.

now the real question is, progmatically, I'm not sure how we know their signing key is bad. We could see if the MAC reveal fails on that particular key then ask them to re-sign in the decryption phase?

Reply:
There's a more comprehensive way to do this, but it's a bit involved.

You can do an operation called a "recover" where if you know the data that was signed (we do) you can back out the original public key (or address) that generated that signature (if it's a valid signature at all).

See this web3 method: https://web3js.readthedocs.io/en/v1.2.11/web3-eth-personal.html#ecrecover. Note: like our other crypto work, this might be tricky to get right, but technically it should be the inverse of this line: https://github.com/UMAprotocol/protocol/blob/4e98402896fe50a5013c8decfaa0f261363ae33c/packages/common/src/Crypto.js#L39. However, because we only need the signature and message, we don't need to ask the user for anythin to do the verification.

Once again, this is something we should add later, since it's not likely in the short term and might be a little tricky to get right.